### PR TITLE
Uplift script handling

### PR DESCRIPTION
**Pull request by RussKie**
## Proposed changes

- Decouple script parsing from the revision grid. This will allow easier and simpler integration of other controls with the script manager (e.g. the left panel or the file tree)
- Extract user script binding to a context menu to an extension method. This will allow re-use of the functionality in other parts of the app (e.g. the left panel or the file tree). This is small step in the same direction as #6348 proposed by @ivangrek. Sadly we stalled on this work.
- Review the code and add multiple validations and add tests.

This work is foundational to allow re-use of user scripts in other parts of the app (e.g. the left panel or the file tree). Specifically I want to allow users to run custom scripts for files in the working tree, e.g. to achieve what was proposed in #6135, or open a file and/or folder in VS Code, or launch a project in a VS.

image

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/Script/ScriptOptionsParser.cs [hunk 1]
```diff
@@ -258,8 +284,18 @@ private static string GetRemotePath(string url)
...
+        private static string ParseScriptArguments([NotNull] string arguments, [NotNull] string option, IWin32Window owner,
+            IScriptHostControl scriptHostControl, IGitModule module, IReadOnlyList<GitRevision> allSelectedRevisions,
+            in IList<IGitRef> selectedTags, in IList<IGitRef> selectedBranches, in IList<IGitRef> selectedLocalBranches,
+            in IList<IGitRef> selectedRemoteBranches, in IList<string> selectedRemotes, GitRevision selectedRevision,
+            in IList<IGitRef> currentTags, in IList<IGitRef> currentBranches, in IList<IGitRef> currentLocalBranches,
+            in IList<IGitRef> currentRemoteBranches, GitRevision currentRevision, string currentRemote)
         {
+            ////if (string.IsNullOrWhiteSpace(arguments) || string.IsNullOrWhiteSpace(option))
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/Script/ScriptOptionsParser.cs:294]: No, it is something to enable. I added it so I won't forget.

File: GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+namespace GitUI.CommandsDialogs
+{
+    public static class UserScriptContextMenuExtensions
+    {
+        /// <summary>
+        ///  Appends user scripts to the <paramref name="contextMenu"/>, or under <paramref name="hostMenuItem"/>,
+        ///  if scripts are marked as <see cref="ScriptInfo.AddToRevisionGridContextMenu"/>.
+        /// </summary>
+        /// <param name="contextMenu">The main context context menu to add user scripts too.</param>
+        /// <param name="hostMenuItem">The context menu item to which to add user scripts marked as <see cref="ScriptInfo.AddToRevisionGridContextMenu"/>.</param>
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:15]: Updated

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+﻿namespace GitUI
+{
+    public sealed partial class RevisionGridControl
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs:3]: @mstv would it be a major pain for you to rebase on top of this, if this goes in first?

[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs:3]: No problem, #7888 needs further time. Setting up the test repo and make FormBrowse load it correctly is a pain.

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        internal enum Commands
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs:5]:
```suggestion
        internal enum Command
```

File: GitUI/Script/ScriptOptionsParser.cs [hunk 5]
```diff
@@ -52,31 +55,41 @@ public sealed class ScriptOptionsParser
...
+        /// <summary>
+        ///  Checks whether <paramref name="arguments"/> contains the supplied <paramref name="option"/>.
+        /// </summary>
+        /// <param name="arguments">The script argument.</param>
+        /// <param name="option">The option.</param>
+        /// <returns><see langword="true"/> if the argument contains the option; otherwise <see langword="false"/>.</returns>
         public static bool Contains([NotNull] string arguments, [NotNull] string option)
         {
+            arguments = arguments ?? throw new ArgumentNullException(nameof(arguments));
+            option = option ?? throw new ArgumentNullException(nameof(option));
```
Comment (mstv) [inline GitUI/Script/ScriptOptionsParser.cs:67]:
```suggestion
            arguments ??= throw new ArgumentNullException(nameof(arguments));
            option ??= throw new ArgumentNullException(nameof(option));
```

File: GitUI/Script/ScriptOptionsParser.cs [hunk 6]
```diff
@@ -52,31 +55,41 @@ public sealed class ScriptOptionsParser
...
         }
 
+        /// <summary>
+        ///  Checks whether <paramref name="option"/> starts with "s".
+        /// </summary>
+        /// <param name="option">The option.</param>
+        /// <returns><see langword="true"/> if the options starts with "s"; otherwise <see langword="false"/>.</returns>
         public static bool DependsOnSelectedRevision([NotNull] string option)
         {
+            option = option ?? throw new ArgumentNullException(nameof(option));
```
Comment (mstv) [inline GitUI/Script/ScriptOptionsParser.cs:79]:
```suggestion
            option ??= throw new ArgumentNullException(nameof(option));
```

File: GitUI/Script/ScriptOptionsParser.cs [hunk 7]
```diff
@@ -52,31 +55,41 @@ public sealed class ScriptOptionsParser
...
 
-        public static (string arguments, bool abort) Parse(string arguments, IGitModule module, IWin32Window owner, RevisionGridControl revisionGrid)
+        public static (string arguments, bool abort) Parse([CanBeNull] string arguments, [NotNull] IGitModule module, IWin32Window owner, IScriptHostControl scriptHostControl)
         {
+            if (string.IsNullOrWhiteSpace(arguments))
+            {
+                return (arguments, abort: false);
+            }
+
+            module = module ?? throw new ArgumentNullException(nameof(module));
```
Comment (mstv) [inline GitUI/Script/ScriptOptionsParser.cs:91]:
```suggestion
            module ??= throw new ArgumentNullException(nameof(module));
```

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs (hunk 4 above)
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.Commands.cs:5]: You can ignore this. I rename it in #7888 anyway.

File: GitUI/Script/ScriptOptionsParser.cs (hunk 7 above)
Comment (RussKie) [reply GitUI/Script/ScriptOptionsParser.cs]: This actually is breaking the build...
image

**Pull request merged**